File: graph/bipartite_graph/bipartite_graph.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

def dfs(now, cor):
    color[now] = cor
    for nxt in graph[now]:
        if color[nxt] == 'W':
            if cor == 'X': dfs(nxt, 'Y')
            else: dfs(nxt, 'X')
        else:
            if color[now] == color[nxt]:
                isBigraph = False
                return
    return


v, e = map(int, input().split())
color = ['W' for _ in range(v+1)]
graph = [[] for _ in range(v+1)]
isBigraph = True
for _ in range(e):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
for i in range(1, v+1):
    if color[i] == 'W': dfs(i, 'X')
    else:
        for nxt in graph[i]:
            if color[i] == color[nxt]:
                isBigraph = False
                break
    if not isBigraph: break
print(1 if isBigraph else 0)

File: graph/bipartite_graph/bipartite_graph.cpp
#include <bits/stdc++.h>
#define fastio cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(false)
#define MAX 20010
typedef long long ll;
using namespace std;

ll v, e, a, b;
vector<vector<ll>> graph(MAX);
char color[MAX];
bool isBigraph;

void init(){
    isBigraph = true;
    for (ll i=0; i<MAX; i++){
        color[i] = 'W';
        graph[i].clear();
    }
}

void dfs(ll now, char cor){
    color[now] = cor;
    for (auto nxt: graph[now]){
        if (color[nxt] == 'W'){
            if (cor == 'X') dfs(nxt, 'Y');
            else dfs(nxt, 'X');
        }
        else{
            if (color[now] == color[nxt]){
                isBigraph = false;
                return;
            }
        }
    }
    return;
}

int main() {
    init();
    cin >> v >> e;
    for (ll i=0; i<e; i++){
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    
    for (ll i=1; i<=v; i++){
        if (color[i] == 'W') dfs(i, 'X');
        else{
            for (auto nxt: graph[i]){
                if (color[i] == color[nxt]){
                    isBigraph = false;
                    break;
                }
            }
        }
        if (!isBigraph) break;
    }
    if (isBigraph) cout << "YES";
    else cout << "NO"; 
}
